### PR TITLE
preferences for audio driver selection in Qt

### PR DESCRIPTION
**Pull request by gewang**
Screenshot (4)

part of the 1.5.0.1 patch
Qt version only and thus primarily affects Windows and Linux
preferences dialog window now has an Audio driver drop-down combo box, populated with available drivers (determined at compile time, with preprocessor: DirectSound, WASAPI, ALSA, Pulse, Jack, CoreAudio
selecting a driver triggers a new probe of available audio I/O devices (which is potentially different depending on the driver) and the Audio output and Audio input drop-down combo boxes are re-populated
settings load and save drivers; default driver is inherited from ChuckAudio

### CODE REVIEW

File: src/qt/mAPreferencesWindow.cpp [hunk 1]
```diff
@@ -83,15 +89,76 @@ const QString mAPreferencesInputChannels = "/VM/InputChannels";
...
+const QString mAPreferencesAudioDriver = "/VM/AudioDriver"; // 1.5.0.1 (ge) added
 
 const QString mAPreferencesEnableNetwork = "/VM/EnableNetworkOTFCommands";
 const QString mAPreferencesEnableAudio = "/VM/EnableAudio";
 
 
+// global symbols expected from RtAudio | 1.5.0.1 (ge) added
+extern "C" const unsigned int rtaudio_num_compiled_apis;
+extern "C" const unsigned int rtaudio_compiled_apis[];
+extern "C" const char * rtaudio_api_names[][2];
```
Comment (spencersalazar) [inline src/qt/mAPreferencesWindow.cpp:101]: seems preferable to have this accessed via `ChuckAudio::` functions

File: src/miniAudicle.h [hunk 2]
```diff
@@ -157,7 +163,9 @@ class miniAudicle
     t_CKBOOL set_named_chugins( std::list< std::string > & chugins );
     t_CKBOOL get_named_chugins( std::list< std::string > & chugins );
     t_CKBOOL add_query_func(t_CKBOOL (*func)(Chuck_Env *));
-    
+    // 1.5.0.1 (ge) added; takes an RtAudio::Api enum
+    t_CKBOOL set_driver(t_CKUINT driver);
```
Comment (spencersalazar) [inline src/miniAudicle.h:167]: style note: here `set_driver` takes a `t_CKUINT` while `probe` takes a `const char *` -- preferable to consolidate to one or the other

also should add a comment where one can find acceptable values for each/either of these -- e.g. `ChuckAudio::getAvailableApis` or something

File: src/qt/ZSettings.h [hunk 3]
```diff
@@ -22,24 +22,44 @@ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
...
-    void set(const QString &key, const QVariant &value);
-    QVariant get(const QString &key, const QVariant &fallback = QVariant());
-    
+    // set value
+    void set( const QString & key, const QVariant & value );
+    // get a value by key
+    QVariant get( const QString & key, const QVariant & fallback = QVariant() );
+
+public:
+    // set defeault
```
Comment (spencersalazar) [inline src/qt/ZSettings.h:56]:
```suggestion
    // set default
```

File: src/qt/mAPreferencesWindow.cpp [hunk 4]
```diff
@@ -116,67 +185,77 @@ void mAPreferencesWindow::configureDefaults()
...
+    // populate the "Audio drivers" ComboBox
+    for( unsigned int i = 0; i < rtaudio_num_compiled_apis; i++ )
+    {
+        // get full driver name
+        snprintf( buffer, sizeof(buffer), "%s", rtaudio_api_names[rtaudio_compiled_apis[i]][1] );
+        // add the item, with the Api enum as the associated data
+        ui->audioDriver->addItem( buffer, rtaudio_compiled_apis[i] );
+        // check if match with driver read in from settings
+        if( driver == rtaudio_compiled_apis[i] ) selectedDriverIndex = i;
+    }
```
Comment (spencersalazar) [inline src/qt/mAPreferencesWindow.cpp:235]: can try something like

```
    struct ChuckAudio::ApiInfo
    {
        t_CKUINT driver;
        std::string name;
        std::string userFriendlyName;
    };

    for( int i = 0; i < ChuckAudio::getNumDrivers(); i++ )
    {
        ChuckAudio::ApiInfo compiledApis = ChuckAudio::getCompiledDriverInfo(i);
        // get full driver name
        // add the item, with the Api enum as the associated data
        ui->audioDriver->addItem( api.userFriendlyName, api.driver );
        // check if match with driver read in from settings
        if( driver == compiledApis.driver ) selectedDriverIndex = i;
    }
```

File: src/qt/mAPreferencesWindow.cpp [hunk 5]
```diff
@@ -300,21 +386,34 @@ void mAPreferencesWindow::ProbeAudioDevices()
...
+// 1.5.0.1 (ge) added
+void mAPreferencesWindow::selectedAudioDriverChanged()
+{
+    // get item data
+    int driver = ui->audioDriver->itemData(ui->audioDriver->currentIndex()).toInt();
+    // trigger probe
+    this->probeAudioDevices(driver);
+    // reset to first items
+    ui->audioOutput->setCurrentIndex( 0 );
+    ui->audioOutput->setCurrentIndex( 0 );
```
Comment (spencersalazar) [inline src/qt/mAPreferencesWindow.cpp:409]:
```suggestion
    ui->audioInput->setCurrentIndex( 0 );
```

File: src/qt/mAPreferencesWindow.cpp [hunk 6]
```diff
@@ -300,21 +386,34 @@ void mAPreferencesWindow::ProbeAudioDevices()
...
+}
+
+
+// 1.5.0.1 (ge) added
+void mAPreferencesWindow::selectedAudioDriverChanged()
+{
+    // get item data
+    int driver = ui->audioDriver->itemData(ui->audioDriver->currentIndex()).toInt();
+    // trigger probe
+    this->probeAudioDevices(driver);
```
Comment (spencersalazar) [inline src/qt/mAPreferencesWindow.cpp:406]: can force reset to default by adding a bool to `probeAudioDevices`

```
void mAPreferencesWindow::probeAudioDevices( int driver, bool resetToDefault )
{
    // ...
    // load available audio I/O interfaces into the pop up menus
    for(i = 0; i < len; i++)
    {
        if(interfaces[i].outputChannels > 0 || interfaces[i].duplexChannels > 0)
        {
            ui->audioOutput->addItem(interfaces[i].name.c_str(), int(i+1));
            if(i + 1 == dac)
                ui->audioOutput->setCurrentIndex(ui->audioOutput->count()-1);
            if((resetToDefault || dac == 0) && interfaces[i].isDefaultOutput)
                ui->audioOutput->setCurrentIndex(ui->audioOutput->count()-1);
        }

        if(interfaces[i].inputChannels > 0 || interfaces[i].duplexChannels > 0)
        {
            ui->audioInput->addItem(interfaces[i].name.c_str(), int(i+1));
            if(i + 1 == adc)
                ui->audioInput->setCurrentIndex(ui->audioInput->count()-1);
            if((resetToDefault || adc == 0) && interfaces[i].isDefaultInput)
                ui->audioInput->setCurrentIndex(ui->audioInput->count()-1);            
        }
    }
```

File: src/qt/mAPreferencesWindow.cpp (hunk 5 above)
Comment (spencersalazar) [reply src/qt/mAPreferencesWindow.cpp:409]: can remove both of these if using `probeAudioDevices` with new `resetToDefault = true`

Comment (gewang) [reply src/qt/mAPreferencesWindow.cpp:409]: done

File: src/qt/mAPreferencesWindow.cpp (hunk 6 above)
Comment (gewang) [reply src/qt/mAPreferencesWindow.cpp:406]: implemented and corrected the input bugs

File: src/qt/ZSettings.h (hunk 3 above)
Comment (gewang) [reply src/qt/ZSettings.h:56]: fixed lol

File: src/qt/mAPreferencesWindow.cpp (hunk 4 above)
Comment (gewang) [reply src/qt/mAPreferencesWindow.cpp:235]: moved direct references to RtAudio out of miniAudicle preferences and accessed through ChuckAudio

File: src/qt/mAPreferencesWindow.cpp (hunk 1 above)
Comment (gewang) [reply src/qt/mAPreferencesWindow.cpp:101]: done

**Review by spencersalazar — APPROVED** [review]

**Pull request merged**